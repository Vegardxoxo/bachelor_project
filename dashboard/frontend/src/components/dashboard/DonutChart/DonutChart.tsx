import React from 'react';
import HelpIcon from '@mui/icons-material/Help';
import {useNavigate} from 'react-router-dom';
import {Box, Card, Stack, Tooltip, Typography} from '@mui/material';
import {Chart, ReactGoogleChartEvent} from 'react-google-charts';


type Props = {
    total_licenses: number,
    active_licenses: number,
    never_used: number,
    unused_licenses: number,
    width?: number;
    height?: number;
    title?: string,
    showInformation?: boolean,
}

/**
 * Returns a DonutChart component displaying the number of active, unused and never used licenses.
 * @param infoBoxData - The data to display in the chart.
 */
function DonutChart(infoBoxData: Props) {

    // Get the function for navigating to different routes
    const navigate = useNavigate();

    // Set default width and height for the chart if not provided
    const {width = 670, height = 425} = infoBoxData;

    // Set default title and display information option for the chart if not provided
    const {title = 'Total oversikt', showInformation = true} = infoBoxData;

    // Define what happens when a slice of the chart is clicked
    const chartEvents: ReactGoogleChartEvent[] = [
        {
            eventName: "select",
            callback: ({chartWrapper}) => {
                const chart = chartWrapper.getChart();
                const selection = chart.getSelection();
                if (selection.length === 1) {
                    const [selectedItem] = selection;
                    const dataTable = chartWrapper.getDataTable();
                    let column = selectedItem.column;
                    if (column === null || column === undefined) {
                        column = 0;
                    }
                    const row = selectedItem.row;
                    console.log("You selected:", {
                        row,
                        column,
                        value: dataTable?.getValue(row, column),
                    });
                    // Navigate to different routes based on the selected slice
                    const value = dataTable?.getValue(row, column)
                    if (value === 'Aktiv') {
                        navigate(`/Totale Lisenser`);

                    }
                    if (value === 'Ledig') {
                        navigate(`/Ledige Lisenser`);
                    }
                    if (value === 'Ubrukt') {
                        navigate(`/Ubrukte Lisenser`);
                    }

                }
            },
        },
    ];

    // Define the data to display in the chart
    const data = [
        ['Type', 'Value'],
        ['Aktiv', infoBoxData.active_licenses ?? 1],
        ['Ledig', infoBoxData.unused_licenses ?? 1],
        ['Ubrukt', infoBoxData.never_used ?? 1],
    ];

    // Define the options for the chart
    const options = {
        pieHole: 0.4,
        legend: 'none',
        pieSliceText: 'percentage',
        colors: ['#80cc9f', '#f9c680', '#f28f8d'],
    };

    // Render the chart component with the data, options and events defined above
    return (
        <Card
            id={"donutChart"}
            data-testid='donutChart'
            sx={{
                height: height,
                width: width,
                mt: 7,
                borderRadius: 5,
            }}
        >
            <Stack direction="column" sx={{height: "100%"}}>
                <Stack direction="row" alignItems="center" sx={{justifyContent: 'space-between'}}>
                    <Typography
                        sx={{
                            textAlign: 'left',
                            fontSize: 27,
                            padding: 2,
                            color: '#002d53',
                            paddingTop: "30px",
                            paddingLeft: "25px",
                            fontFamily: 'Source Sans Pro,sans-serif',
                        }}
                    >
                        {title}
                    </Typography>
                    {/* An optional help icon displaying information about the chart */}
                    {showInformation &&
                        <Tooltip
                            title={
                                <h2 style={{fontSize: 15, fontWeight: 'lighter'}}>
                                    Diagrammet viser en oversikt over hvor mange aktive, ubrukte og ledige
                                    lisenser som eies i enheten.
                                </h2>
                            }
                            placement="top"
                            arrow
                        >
                            <HelpIcon
                                sx={{color: 'grey', fontSize: 25, marginRight: 2, marginTop: 2}}
                                data-testid="donutchartHelpIcon"
                            ></HelpIcon>
                        </Tooltip>
                    }
                </Stack>
                {/* Additional information about what the different colours mean in the donut chart */}
                <Stack direction="row" sx={{paddingLeft: 7, height: "100%", width: "100%", alignItems: 'center'}}>
                    <Stack spacing={5.2} sx={{paddingTop: 7.5, paddingLeft: 3, height: "100%"}}>
                        <Box sx={{width: 18, height: 18, backgroundColor: '#80cc9f'}}/>
                        <Box sx={{width: 18, height: 18, backgroundColor: '#f9c680'}}/>
                        <Box sx={{width: 18, height: 18, backgroundColor: '#f28f8d'}}/>
                    </Stack>
                    <Stack
                        spacing={4}
                        sx={{
                            paddingTop: 7,
                            paddingLeft: 3,
                            paddingRight: -50,
                            color: '#002d53',
                            fontStyle: 'Source Sans Pro,sans-serif',
                            height: "100%",
                            fontSize: "30pt",
                        }}
                    >
                        <Typography sx={{fontFamily: 'Source Sans Pro,sans-serif', fontSize: "14pt"}}>Aktiv</Typography>
                        <Typography sx={{fontFamily: 'Source Sans Pro,sans-serif', fontSize: "14pt"}}>Ledig</Typography>
                        <Typography sx={{fontFamily: 'Source Sans Pro,sans-serif', fontSize: "14pt"}}>Ubrukt</Typography>
                    </Stack>

                    <Chart
                        chartType="PieChart"
                        width="100%"
                        height="100%"
                        data={data}
                        options={options}
                        chartEvents={chartEvents}
                        style={{cursor: 'pointer'}}
                    />

                </Stack>
            </Stack>
        </Card>
    )


}

export default DonutChart;