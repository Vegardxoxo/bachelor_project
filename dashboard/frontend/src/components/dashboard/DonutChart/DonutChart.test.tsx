import {act, cleanup, fireEvent, render, screen} from '@testing-library/react';
import DonutChart from './DonutChart';
import '@testing-library/jest-dom/extend-expect';
import {MemoryRouter} from 'react-router-dom';
import renderer from "react-test-renderer";
import React from "react";

/* Mockdata for infoboxdata */
const infoBoxData = {
    total_licenses: 70,
    active_licenses: 40,
    never_used: 10,
    unused_licenses: 30,
    available_licenses: 40
}

jest.mock('../../../api/calls', () => ({
    fetchInfoBoxData: () => Promise.resolve(infoBoxData)
}));

describe('Testing render, DonutChart', () => {

    beforeEach(async () => {
        await act(async () => {
            render(
                <MemoryRouter>
                    <DonutChart never_used={infoBoxData.never_used}
                                total_licenses={infoBoxData.total_licenses}
                                unused_licenses={infoBoxData.unused_licenses}
                                active_licenses={infoBoxData.active_licenses}/>
                </MemoryRouter>
            );
        });
    });

    afterEach(() => {
        localStorage.clear();
        cleanup();
    });

    it("renders donutchart without crashing", async () => {
        expect(screen.getByTestId('donutChart')).toBeInTheDocument();
        expect(screen.getByText('Aktiv')).toBeInTheDocument();
        expect(screen.getByText('Ledig')).toBeInTheDocument();
        expect(screen.getByText('Ubrukt')).toBeInTheDocument();
    });

    it('matches snapshot', async () => {
        const testRenderer = renderer.create(
            <MemoryRouter>
                <DonutChart never_used={infoBoxData.never_used}
                            total_licenses={infoBoxData.total_licenses}
                            unused_licenses={infoBoxData.unused_licenses}
                            active_licenses={infoBoxData.active_licenses}/>
            </MemoryRouter>
        );
        expect(testRenderer.toJSON()).toMatchSnapshot();
    })


    it('renders tooltip when hovering over HelpIcon', async () => {
        fireEvent.mouseEnter(screen.getByTestId('donutchartHelpIcon'))
        await screen.findByRole(/tooltip/);
        expect(screen.getByRole(/tooltip/)).toBeInTheDocument();
    });


});


