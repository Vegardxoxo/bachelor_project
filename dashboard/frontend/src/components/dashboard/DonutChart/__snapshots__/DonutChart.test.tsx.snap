// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing render, DonutChart matches snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-qypn6k-MuiPaper-root-MuiCard-root"
  data-testid="donutChart"
  id="donutChart"
>
  <div
    className="css-1rzuxdl-MuiStack-root"
  >
    <div
      className="css-1jv2z0n-MuiStack-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-m4473c-MuiTypography-root"
      >
        Total oversikt
      </p>
      <svg
        aria-hidden={true}
        aria-label={null}
        aria-labelledby={null}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-p51dmq-MuiSvgIcon-root"
        data-mui-internal-clone-element={true}
        data-testid="donutchartHelpIcon"
        focusable="false"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
        />
      </svg>
    </div>
    <div
      className="css-15hav22-MuiStack-root"
    >
      <div
        className="css-hqhp0o-MuiStack-root"
      >
        <div
          className="MuiBox-root css-vsxoph"
        />
        <div
          className="MuiBox-root css-4kogbh"
        />
        <div
          className="MuiBox-root css-1tkh0sf"
        />
      </div>
      <div
        className="css-7ivaql-MuiStack-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-1esljnb-MuiTypography-root"
        >
          Aktiv
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1 css-1esljnb-MuiTypography-root"
        >
          Ledig
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1 css-1esljnb-MuiTypography-root"
        >
          Ubrukt
        </p>
      </div>
    </div>
  </div>
</div>
`;
